### ISSUE TITLE
method-probing machinery can false-positive

### ISSUE THREAD

**Issue by zackmdavis** [rust-lang/rust#46459]
[While working on improving method suggestions](https://github.com/rust-lang/rust/issues/42929#issuecomment-345914556), it was discovered that the return value of `probe_for_return_type`—at least, a slightly altered version thereof that passes `ProbeScope::AllTraits` to `probe_op` rather than [`ProbeScope::TraitsInScope` (as on master at issue-filing time)](https://github.com/rust-lang/rust/blob/088f328e/src/librustc_typeck/check/method/probe.rs#L186-L195)—could include methods that don't actually return the desired type. Most notably, this happened with `core::convert::Into::into`.

The root cause _seems_ to be that when we [try to resolve type variables](https://github.com/rust-lang/rust/blob/088f328e/src/librustc_typeck/check/method/probe.rs#L1084), we might [find that the return type of the trait method still needs to be resolved](https://github.com/rust-lang/rust/blob/088f328e/src/librustc/infer/mod.rs#L1282-L1284), and that unresolved type variable is [considered to fulfill the required return type](https://github.com/rust-lang/rust/blob/78fcf3388/src/librustc_typeck/check/method/probe.rs#L1089)?? (I regret my difficulty to provide a better report than this.)

**Comment by zackmdavis** [comment]
To hopefully make this slightly more concrete: on my branch, this led to us suggesting adding `.into()` to fix

```rust
fn main() {
    let x: usize = String::new();
}
```
which does not work.
